### DISCUSSION TITLE
Extension approach proposal

### DISCUSSION THREAD

**Discussion by DGuhr** [keycloak/keycloak#13181] [Archive]
I've just posted this proposal on the quarkus-dev mailing list: https://groups.google.com/g/quarkus-dev/c/_rOQbUkQbL8/m/mV559sKkCAAJ

Content duplicated here: 

...
**Problem statement**
A good chunk of Keycloaks value comes from its extensibility and its ability to integrate well in various environments, from on-prem to k8s, from AWS to OpenShift, from Oracle to Postgres.

Currently, Keycloak is very opinionated about the extensions available out of the box in its distribution for various reasons listed below. This is needed, but makes life hard for developers and users of Keycloak when they want to use additional extensions that are already available in the Quarkus universe.

Keycloak itself is a quarkus extension under the covers, and uses the mutable-jar approach to provide great flexibility when addressing the different environments it runs in. 
Keycloak relies on the re-augmentation step to provide an optimized runtime with fast startup and low memory footprint. Mutable-jar applications do not allow extensions other than those available at compile time to be installed when re-augmenting the application.

Keycloak as a layered product can not include Quarkus extensions OOTB, other than those that are strictly necessary to support its core capabilities. Otherwise, this can cause:

- too big distribution size
- High maintenance cost for the core team
- Bad UX / Hard to choose what extensions should be available by default as they are environment / use case specific (e.g.: you might want or not a JAX-RS Client, centralized logging, …)
- Higher risk of e.g. CVEs -> more dependencies are included and available from the distribution even when the environment Keycloak runs in only needs a subset of, for example, database drivers.

**Use cases / stories**
_This section should give you a general idea of what we have in mind mid- to long term, trying not to go into technical solution details but exploring the problem space. It is not considered complete and likely to be extended/changed in the future. See “Action points” for more specific points we'd love to hear your thoughts about._

**As a** Keycloak developer I want to be able to leverage Quarkus extensions and use them when implementing my custom providers so that I can take advantage of the available extensions from the Quarkus ecosystem.

For a better developer experience, developers should be able to:

- Choose an extension from Quarkus ecosystem
- Write their custom providers using the extensions
- Package their custom providers and deploy to the server 
- Run re-augmentation to create a new “closed world assumption” based on the chosen extensions and providers
- Have their custom providers and Quarkus extensions installed and available at runtime

**As a** Keycloak developer I want to be able to leverage Quarkus extensions and use them when implementing my custom providers ( == "custom keycloak extensions") so that I can take advantage of the available extensions from the Quarkus ecosystem.

**As a** platform team member of a “customer”, I want to tailor our platforms Keycloak component specific to my operational environment, with the least complexity possible, so that the risk of being affected by e.g. CVEs from dependencies not used in my environment is lowered and compliance is fullfilled (some historical context: https://github.com/keycloak/keycloak-community/pull/328/files#r735664241.. )
- I want one reliable, secure “marketplace” to go to to tailor Keycloak to my specific needs
- I want to know what is “supported upstream by the core project” and what is not.
  - Example: The marketplace offers extensions for different databases, where some are community supported, meaning that there is no direct CI run to test Keycloaks behaviour when using these databases, some are “core extensions” (see story below).
  - Example 2: The marketplace offers extensions for being able to integrate Keycloak with different observability solutions. 
  - Example 3: The marketplace offers extensions for being able to connect Keycloak to eventing systems.
  - Example 4: The marketplace offers extensions for being able to integrate Keycloak with different centralized logging solutions. 
…
- I want to download a packaged “distribution” containing the tailored Keycloak+extensions package. (Maybe even pre-optimized as a fast-jar instead of a mutable-jar)
- I want to download/create a git repository of a developer boilerplate ready for extending Keycloak with my internal extensions. 
- (Later:) I want to create the tailored Keycloak distribution as a docker image directly in e.g. my orgs [quay.io](http://quay.io/) registry account
- (Even later: ) I want my tailored “closed world” Keycloak to be downloaded / uploaded to a container registry from the marketplace as a native image.

**As a** maintainer of Keycloak core, I want to strip away some features of the current monolithic distribution into available, supported “core extensions”. The “software formerly known as distribution” will still be there, but will consist of a lightweight “core Keycloak” codebase, handling AuthN and AuthZ (our core domain), and a set of supported core extensions packaged into it. Our pipelines will still test this full package, thus making it “upstream supported”, but we will have the flexibility to remove stuff from “core” that is not strictly necessary there, to reduce maintenance cost and increase focus on our core capabilities.
- E.g. some social identity providers like Twitter, ...
- E.g. support for LDAP connections (to not necessarily have JNDI enabled for our users, that is needed to communicate with LDAP and enabled by default atm)
… 

**As a** platform team member of a “customer”, I want to be able to leverage the huge quarkiverse ecosystem by installing my needed extensions, to tailor Keycloak specific to my platform's Keycloak component needs.
- E.g. by invoking `kc extensions install <quarkiverse-extension/MAR-with-metadata-for-compatibility>` followed by a `kc build` to activate the extension

**As a** platform team member of a “customer”, I want to be able to search the huge Quarkiverse ecosystem, to find the available extensions together with their support status for the current Keycloak version.

- E.g. by invoking `kc extensions search <query>`

**The current state**
People are actually able to use some Quarkiverse extensions in Keycloak, but in an experimental, non-convenient and unsupported way: 
- Download the `<extension>-deployment.jar` and the `<extension>.jar` and **all the transient jars** the extensions use
- Move them to Keycloaks `providers` folder
- Run a `kc.sh build` (this command invokes a re-augmentation of the mutable jar)
- Start Keycloak. The extension is registered, a new closed world assumption created. Or not, when anything goes wrong (e.g. quarkiverse.logging vs quarkus.logging)

**Drawbacks:**
There is no ruleset/metadata to check if a (set of) extensions are compatible with each other and with the extensions Keycloak uses OOTB (e.g. quarkiverse-logging-json and quarkus-logging-json can not be used together). 
The whole process is experimental / non-idiomatic, the augmentation guide states “ Augmentation can only use extensions that were present during compile time.” 

**Potential Solution**
One possible solution to make it possible to add additional extensions to a mutable application like Keycloak may be to introduce the concept of a MAR (Mutable Application Archive) that relies on the JAR format and adds a well-defined, structured “Mutable Application extension” contract, that describes the extensions that should be installed during re-augmentation and the user code using these extensions.

**Action points:**
In order to achieve this vision, we would love to talk with you about:
- Your general feedback of the viability of this idea, basically it’s kind of a “layered framework approach”. 
- The MAR contract for the necessary metadata to integrate quarkus extensions into Keycloak / layered products 
- Figure out how to dynamically enable or disable extensions with mutable jars
- …? We want your feedback :)

**Comment by alechenninger** [comment]
All of this sounds pretty great. If I had to rank what was most important to us:

1. Supported ability to add extensions (must have)
2. Supported way to integrate and reuse Quarkus extensions (our custom extensions will likely need common libraries that Quarkus already has integrations with)
3. Ability to remove components we don't need (very nice for lowering CVE surface area, footprint, etc.)

For our team, it's probably not too important to have pre-built tailor-made distributions–we're comfortable building these ourselves. Similarly, a built-in way to search for Quarkus extensions is not a huge benefit, as we'd probably just be using Google etc. anyway :-).

But in general, I really like the idea of treating extensions as more first class and taking advantage of Quarkus platform and ecosystem. I appreciate the recognition of value that customizability has. While you can go overboard with it (heh),  a core tenet of being "open" isn't just "I can see the source" but "I can extend and customize for my needs."

**Reply by bs-matil** [reply]
I agree ad would add that extension search may kicks in if the amount of extensions grows + the cli also has some way to directly add the extension to the current project

**Comment by danielFesenmeyer** [comment]
Sounds great.

What I would appreciate in addition to the improvements mentioned above: the ability to enable/disable (preview) features and SPI implementations at startup, without re-augmentation - because re-augmentation takes quite long in case of many SPI implementations.
See:
- https://github.com/keycloak/keycloak/discussions/10886
- https://github.com/keycloak/keycloak/discussions/10986

**Comment by MariuszCwikla** [comment]
It would be great to have such support for Quarkus extensions in Keycloak.

I managed to do some demo with Quarkus Kafka.
More or less here is the procedure

- create a maven module that adds dependency to `keycloak-quarkus-server`
- you need to tweak some dependencies, e.g. add `keycloak-common` explicitely to force upgrade (quarkus and keycloak conflicts; quarkus provides version 16.x); liquibase needs to be downgraded too
- add your extensions (I have added `quarkus-smallrye-reactive-messaging-kafka`)
- you need to use `jandex-maven-plugin`

The main trick is to use `quarkus-extension-processor` that will build Quarkus App with all Keycloak libraries and will generate other Quarkus artifacts (`lib/deployment/appmodel.dat`, `lib/lib/deployment/deployment-class-path.dat`).
Then just copy contents of `target/lib` into docker container. More or less the Dockerfile looks like this:

```
FROM quay.io/keycloak/keycloak:20.0.3

RUN rm -rf /opt/keycloak/lib/lib
COPY --chown=keycloak target/lib/ /opt/keycloak/lib/
```

If anyone wants I can share the demo.

I haven't tested it fully though. There are few other issues, e.g. I noticed that it adds bit older version of Infinispan. You would need to update those dependencies manually. I guess few other libraries might introduce similar problem.

Anyway it would be great to have official support for this.

**Reply by vanti1980** [reply]
could you pls share it?
